Google Traduction with Deep Translator

jjxav opened the pull request:
Add the google traduction api with [deep-translator](https://pypi.org/project/deep-translator/#deepltranslator)

jjxav committed:
Implementation of google traduction api

PladsElsker commented on poetry.lock:
```diff
@@ -1801,6 +1850,35 @@ files = [
+[[package]]
+name = "torch"
+version = "2.1.0"
+description = "Tensors and Dynamic neural networks in Python with strong GPU acceleration"
```
Can we remove pytorch from the poetry file? We already have it installed with the `poe torch-cuda` command.

PladsElsker replied:
Also, this is going to install the cpu version, we probably don't want that.

lbaret replied:
It may be a good idea ! But anyway if you install it and run `poe torch-cuda` then it will be fine. So as you want.

jjxav replied:
Torch is a dependency of the packages manga-ocr that was part of #14. When it is removed from the lock, an error is thrown and so I will not remove it for our future install to be successful.

![image](https://github.com/cia-ulaval/MangaAutoTranslator/assets/111534376/c4bf89d1-0a92-49a6-a640-71df7fc6c009)

PladsElsker replied:
That's because you also need to remove the line 747 `torch = ">=1.0"`. I don't know why this is in main, I thought I removed it.

PladsElsker replied:
So if you remove the explicit dependency of torch + pytorch, it's not gonna flag it as an error (at least it didn't when I did it on my side).

PladsElsker replied:
I confirm that I did not remove the line 747, and that I should have.

PladsElsker commented on manga_auto_translator/translation.py:
```diff
@@ -0,0 +1,38 @@
+        
+    def run(self, scans: Sequence[Scan], lang_from: str, lang_to: str) -> None:
+        from deep_translator import GoogleTranslator
+        self.tranlator = GoogleTranslator(source=lang_from, target=lang_to)
```
I think the creation of the translation model should be moved to the `__init__` function. Same with the arguments.

PladsElsker commented on manga_auto_translator/pipeline.py:
```diff
@@ -1,12 +1,16 @@
 
 class TranslationPipeline:
-    def __init__(self, scans: Sequence[Scan], ocr_strategy: OcrStrategy) -> None:
+    def __init__(self, scans: Sequence[Scan], lang_from: str, lang_to: str, ocr_strategy: OcrStrategy, translation_strategy: TranslationStrategy) -> None:
```
We can remove the 2 language arguments here if they are passed to the constructor of the `TranslationStrategy`

lbaret replied:
I second this

PladsElsker reviewed (changes requested):
Good job overall! 💯

PladsElsker commented on manga_auto_translator/pipeline.py:
```diff
@@ -25,6 +29,7 @@ def ocr(self):
 
     def translation(self):
+        self.translation_strategy.run(self.scans, self.lang_from, self.lang_to)
         pass
```
Small detail, but the `pass` keyword can be removed now that there is actual code in the function.

John-WL replied:
Also, we should probably add typing in the function parameters.

jjxav replied:
Fixed

lbaret commented on manga_auto_translator/main.py:
```diff
@@ -23,12 +24,31 @@
     scans = ScanIOManager.load_scans(input_path)
 
     ocr_strategy = OcrStrategyFactory(strategy=ocr).create()
+    translation_strategy = TraductionStrategyFactory(strategy=translation_api).create()
```
This is great and I'm doing the same thing.
The way you call it is fine, but can we find a more elegant way to call `create()` instead of instantiating then directly call it ? Can we consider the option to create static method and put `translation_api` as argument for `create()` method ?

PladsElsker replied:
I agree with you. I think this would be better. 
I was thinking of making a follow up PR to suggest this because it's also implemented like that for the OCR.

lbaret commented on manga_auto_translator/translation.py:
```diff
@@ -0,0 +1,38 @@
+        print('Loading GoogleTraductionTranslation...')
+        
+    def run(self, scans: Sequence[Scan], lang_from: str, lang_to: str) -> None:
+        from deep_translator import GoogleTranslator
```
I am not a huge fan of imports in method scope, for testing purpose it may be better to pass it as argument.

lbaret replied:
Maybe you have a reason to do like this, and I'm interested to know. It's not the first time I see this case, and some of greatest libraries used in ML are using this trick.

PladsElsker replied:
I don't know about the `GoogleTranslator` library, but `MangaOCR` takes literally forever to load. If someone uses another strategy, they would still need to wait for the library to load, even if they don't use it. 

Importing from the function itself means that you only import it if you use the strategy. 
So it's mostly to prevent loading unecessary modules.

PladsElsker replied:
So if the loading time for the `GoogleTranslator` library is fast enough, the import should probably be placed with the other imports at the top of the file. Otherwise, I think it's fine to keep it here.

PladsElsker replied:
Maybe we could create an anotation?
Something like:
```py
@module_argument(from='deep_translator', import='GoogleTranslator')
def run(self, scans: Sequence[Scan], lang_from: str, lang_to: str, GoogleTranslator: Module) -> None:
    ...
```

Not sure if `Module` is the right type, but the idea is there.

jjxav replied:
GoogleTranslator does not take long to import. So I've gone with the import at the top of the file.

For the decorator, we could do it, but for the translation in particular none of the other classes for the other api take long to import, so I do not see any benefit with the decorator.

jjxav committed:
Remove Pass keyword

jjxav committed:
Reformat class TranslationFactory & GoogleTraduction

- TranslationFactory pass Lang options to created instance of the class
-  TranslationFactory remove the init parameter and instead use a static function to pass the strategy name

jjxav committed:
Remove poetry from lock
